lib/docker.ts

```diff
@@ -132,69 +174,99 @@ class Docker {
-            if (!Reflect.has(result.md5, filename)) {
-              result.md5[filename] = extracted.layers[layerName].md5[filename];
-            }
-          }
-        }
-      }
+  /**
+   * Attempt to perform a static scan
+   * @param searchActions
+   */
+  public async maybeStaticScan(searchActions: SearchAction[]) {
+    const size = await this.sizeSafe();
+    if (!size || size > this.GetStaticScanSizeLimit()) {
+      return {};
+    }
+    return await this.extractAndCache(searchActions);
+  }
+
+  /**
+   * Get file product that was previously retrieved and cached or by using
+   *  runtime method after applying the specified callback
+   * @param filename name of file to retrieve its associated string product
+   * @param callbacks optional array of callbacks to call when runtime method
+   *  is used to retrieve the file
+   * @returns map of file products by callback name
+   */
+  public async getFilesProducts(
+    filename: string,
+    callbacks?: SearchActionCallback[],
+  ): Promise<{ [key: string]: string | Buffer }> {
+    if (Reflect.has(this.searchActionProducts, filename)) {
+      return this.searchActionProducts[filename];
+    }
+    const content = (await this.catSafe(filename)).stdout;
+    if (!callbacks) {
+      return { "": content };
+    }
+    const result: { [key: string]: string | Buffer } = {};
+    for (const callback of callbacks) {
+      result[callback.name] = callback.call(Buffer.from(content, "utf8"));
     }
     return result;
   }
 
   /**
-   * Saves the docker image as a TAR file to a temporary location and extract
-   * the specified files from it.
-   * @param txtPatterns list of plain text key files paths patterns to extract and return as strings
-   * @param md5Patterns list of binary key files paths patterns to extract and return their MD5 sum
-   * @return list of plain text files and list of MD5 sums
+   * Convenience function the single product of a single file
+   * @param filename name of file to retrieve its associated single product
+   * @param callback optional callback to call when runtime method is used
+   *  to retrieve the file
+   * @returns file product, the only one
    */
-  public async extract(
-    txtPatterns: string[],
-    md5Patterns: string[] = [],
-  ): Promise<ExtractedKeyFiles> {
-    return this.save(async (err, imageTarPath) => {
-      if (err) {
-        throw err;
-      }
-      return await this.analyze(imageTarPath, txtPatterns, md5Patterns);
-    });
+  public async getFileProduct(
+    filename: string,
+    callback?: SearchActionCallback,
+  ): Promise<string> {
+    const result = await this.getFilesProducts(
+      filename,
+      callback ? [callback] : undefined,
+    );
+    const length = Object.keys(result).length;
+    if (length > 1) {
+      throw new Error(`File product ambiguity, ${length}`);
+    }
+    // return the first and only product of the file
+    return result[Object.keys(result)[0]].toString("utf8");
```

Comment:
Why do you need the `utf8`? isn't that function support content that is not only txt?

Reply:
let's do as we talked - create a different function for `txt` only so the call to `getFilesProducts` will be separate from `catSafe`

Reply:
Done